core/src/main/kotlin/org/jetbrains/research/testspark/core/generation/llm/LLMWithFeedbackCycle.kt

```diff
@@ -60,226 +44,146 @@ class LLMWithFeedbackCycle(
-                }
-                break
-            }
 
-            // clearing test assembler's collected text on the previous attempts
-            testsAssembler.clear()
-            val response: Result<TestSuiteGeneratedByLLM, TestSparkError> =
-                requestManager.request(
-                    language = language,
-                    prompt = nextPromptMessage,
-                    indicator = indicator,
-                    packageName = packageName,
-                    testsAssembler = testsAssembler,
-                    isUserFeedback = false,
-                    errorMonitor,
-                )
-
-            // Process stopped checking
-            if (indicator.isCanceled()) throw ProcessCancelledException(module = TestSparkModule.Llm())
-
-            when (response) {
-                is Result.Success -> {
-                    log.info { "Test suite generated successfully: ${response.data}" }
-                    // check that there are some test cases generated
-                    if (response.data.testCases.isEmpty()) {
-                        onWarningCallback?.invoke(LlmError.EmptyLlmResponse)
-
-                        nextPromptMessage =
-                            "You have provided an empty answer! Please answer my previous question with the same formats."
+    fun run(): Flow<Result<TestSuiteGeneratedByLLM>> =
+        flow {
+            var iteration = 0
+            var nextPromptMessage = initialPromptMessage
+            val generatedTestSuites: MutableList<TestSuiteGeneratedByLLM> = mutableListOf()
+
+            while (iteration < requestsCountThreshold) {
+                iteration++
+                log.info { "Iteration #$iteration of feedback cycle" }
+
+                // ensure the testsAssembler is empty before each iteration
+                testsAssembler.clear()
+
+                val chunks: Flow<Result<String>> =
+                    chatSessionManager.request(
+                        prompt = nextPromptMessage,
+                        isUserFeedback = false,
+                    )
+                val testSuiteResult: Result<TestSuiteGeneratedByLLM> = chunks.collectChunks(testsAssembler)
+
+                when (testSuiteResult) {
+                    is Result.Success -> log.info { "Test suite generated successfully: ${testSuiteResult.data}" }
+
+                    is Result.Failure -> {
+                        log.info { "Cannot parse a test suite from the LLM response. LLM response: '$testSuiteResult'" }
+                        emit(testSuiteResult)
+                        nextPromptMessage = generatePromptMessage(testSuiteResult.error) ?: break
+
+                        /**
+                         * The current attempt does not count as a failure since it was rejected due to the prompt size
+                         * exceeding the threshold
+                         */
+                        if (testSuiteResult.error is LlmError.PromptTooLong) iteration--
```

Comment:
The new logic keeps the same kind of bug as #483
If the prompt is too long, the chatHistory is not cleared. So, the next request will send 2 user messages with the original and "reduced" prompts. The error will be the same.